File: sdks/php/src/Api/Webhooks.php
<?php

declare(strict_types=1);

namespace Apollo\Deploy\Signal\Sdk\Api;

use Apollo\Deploy\Signal\Sdk\SDKError;
use Apollo\Deploy\Signal\Sdk\SSEEvent;
use Apollo\Deploy\Signal\Sdk\Transport;
use Apollo\Deploy\Signal\Sdk\CreateWebhookBody;
use Apollo\Deploy\Signal\Sdk\UpdateWebhookBody;
use Apollo\Deploy\Signal\Sdk\WebhookCreateResponse;
use Apollo\Deploy\Signal\Sdk\WebhookDeliveryPageResponse;
use Apollo\Deploy\Signal\Sdk\WebhookDeliveryResponse;
use Apollo\Deploy\Signal\Sdk\WebhookPageResponse;
use Apollo\Deploy\Signal\Sdk\WebhookResponse;
use Psr\Http\Message\ResponseInterface;

/** WebhooksAPI domain operations. */
class Webhooks
{
    public function __construct(private Transport $transport)
    {
    }

    /**
     * @param string $projectId
     * @return WebhookPageResponse
     * @throws SDKError
     */
    public function listWebhooks(
        string $projectId,
    ): WebhookPageResponse
    {
        $path = '/v1/projects/${projectId}/webhooks';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'GET',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(WebhookPageResponse::class, 'fromArray')) {
            return WebhookPageResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $projectId
     * @param string $endpointId
     * @return WebhookResponse
     * @throws SDKError
     */
    public function getWebhook(
        string $projectId,
        string $endpointId,
    ): WebhookResponse
    {
        $path = '/v1/projects/${projectId}/webhooks/${endpointId}';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);
        $path = str_replace('${endpointId}', rawurlencode(Transport::wireValue($endpointId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'GET',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(WebhookResponse::class, 'fromArray')) {
            return WebhookResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $projectId
     * @param string $endpointId
     * @return WebhookDeliveryPageResponse
     * @throws SDKError
     */
    public function listWebhookDeliveries(
        string $projectId,
        string $endpointId,
    ): WebhookDeliveryPageResponse
    {
        $path = '/v1/projects/${projectId}/webhooks/${endpointId}/deliveries';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);
        $path = str_replace('${endpointId}', rawurlencode(Transport::wireValue($endpointId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'GET',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(WebhookDeliveryPageResponse::class, 'fromArray')) {
            return WebhookDeliveryPageResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $projectId
     * @param string $endpointId
     * @param string $deliveryId
     * @return WebhookDeliveryResponse
     * @throws SDKError
     */
    public function getWebhookDelivery(
        string $projectId,
        string $endpointId,
        string $deliveryId,
    ): WebhookDeliveryResponse
    {
        $path = '/v1/projects/${projectId}/webhooks/${endpointId}/deliveries/${deliveryId}';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);
        $path = str_replace('${endpointId}', rawurlencode(Transport::wireValue($endpointId)), $path);
        $path = str_replace('${deliveryId}', rawurlencode(Transport::wireValue($deliveryId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'GET',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(WebhookDeliveryResponse::class, 'fromArray')) {
            return WebhookDeliveryResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $projectId
     * @param CreateWebhookBody $input
     * @return WebhookCreateResponse
     * @throws SDKError
     */
    public function createWebhook(
        string $projectId,
        CreateWebhookBody $input,
    ): WebhookCreateResponse
    {
        $path = '/v1/projects/${projectId}/webhooks';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'POST',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];
        if ($input !== null) {
            $request['body'] = $input;
            $request['contentType'] = 'application/json';
        }

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(WebhookCreateResponse::class, 'fromArray')) {
            return WebhookCreateResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $projectId
     * @param string $endpointId
     * @param UpdateWebhookBody $input
     * @return WebhookResponse
     * @throws SDKError
     */
    public function updateWebhook(
        string $projectId,
        string $endpointId,
        UpdateWebhookBody $input,
    ): WebhookResponse
    {
        $path = '/v1/projects/${projectId}/webhooks/${endpointId}';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);
        $path = str_replace('${endpointId}', rawurlencode(Transport::wireValue($endpointId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'PATCH',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];
        if ($input !== null) {
            $request['body'] = $input;
            $request['contentType'] = 'application/json';
        }

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(WebhookResponse::class, 'fromArray')) {
            return WebhookResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $projectId
     * @param string $endpointId
     * @return void
     * @throws SDKError
     */
    public function deleteWebhook(
        string $projectId,
        string $endpointId,
    ): void
    {
        $path = '/v1/projects/${projectId}/webhooks/${endpointId}';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);
        $path = str_replace('${endpointId}', rawurlencode(Transport::wireValue($endpointId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'DELETE',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $this->transport->readResponseBody($response);
        return;
    }

    /**
     * @param string $projectId
     * @param string $endpointId
     * @return WebhookDeliveryResponse
     * @throws SDKError
     */
    public function testWebhook(
        string $projectId,
        string $endpointId,
    ): WebhookDeliveryResponse
    {
        $path = '/v1/projects/${projectId}/webhooks/${endpointId}/test';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);
        $path = str_replace('${endpointId}', rawurlencode(Transport::wireValue($endpointId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'POST',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(WebhookDeliveryResponse::class, 'fromArray')) {
            return WebhookDeliveryResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $projectId
     * @param string $endpointId
     * @param string $deliveryId
     * @return void
     * @throws SDKError
     */
    public function replayWebhookDelivery(
        string $projectId,
        string $endpointId,
        string $deliveryId,
    ): void
    {
        $path = '/v1/projects/${projectId}/webhooks/${endpointId}/replay/${deliveryId}';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);
        $path = str_replace('${endpointId}', rawurlencode(Transport::wireValue($endpointId)), $path);
        $path = str_replace('${deliveryId}', rawurlencode(Transport::wireValue($deliveryId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'POST',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $this->transport->readResponseBody($response);
        return;
    }

    private function decodeResponse(ResponseInterface $response): mixed
    {
        $contents = $this->transport->readResponseBody($response);
        if (trim($contents) === '') {
            $requestId = $response->getHeaderLine('x-request-id')
                ?: $response->getHeaderLine('request-id')
                ?: null;
            throw new SDKError(
                message: 'Expected a response body for a non-void operation',
                status: $response->getStatusCode(),
                code: 'RESPONSE_BODY_MISSING',
                requestId: $requestId,
            );
        }

        try {
            return json_decode($contents, true, 512, JSON_THROW_ON_ERROR);
        } catch (\JsonException) {
            return $contents;
        }
    }
}

File: sdks/rust/src/api/webhooks.rs
use crate::error::SdkError;
use crate::transport::Transport;
use reqwest::Method;
use serde::Serialize;
use crate::types::{ CreateWebhookBody, UpdateWebhookBody, WebhookCreateResponse, WebhookDeliveryPageResponse, WebhookDeliveryResponse, WebhookPageResponse, WebhookResponse };

/// WebhooksAPI API operations.
pub struct WebhooksAPI {
    transport: Transport,
}

impl WebhooksAPI {
    pub(crate) fn new(transport: Transport) -> Self {
        Self { transport }
    }

    pub async fn list_webhooks(
        &self,
        project_id: String,
    ) -> Result<WebhookPageResponse, SdkError> {
        let mut path = "/v1/projects/${projectId}/webhooks".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<WebhookPageResponse>(
            Method::GET,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await
    }

    pub async fn get_webhook(
        &self,
        project_id: String,
        endpoint_id: String,
    ) -> Result<WebhookResponse, SdkError> {
        let mut path = "/v1/projects/${projectId}/webhooks/${endpointId}".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));
        path = path.replace("${endpointId}", &encode_path(&endpoint_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<WebhookResponse>(
            Method::GET,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await
    }

    pub async fn list_webhook_deliveries(
        &self,
        project_id: String,
        endpoint_id: String,
    ) -> Result<WebhookDeliveryPageResponse, SdkError> {
        let mut path = "/v1/projects/${projectId}/webhooks/${endpointId}/deliveries".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));
        path = path.replace("${endpointId}", &encode_path(&endpoint_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<WebhookDeliveryPageResponse>(
            Method::GET,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await
    }

    pub async fn get_webhook_delivery(
        &self,
        project_id: String,
        endpoint_id: String,
        delivery_id: String,
    ) -> Result<WebhookDeliveryResponse, SdkError> {
        let mut path = "/v1/projects/${projectId}/webhooks/${endpointId}/deliveries/${deliveryId}".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));
        path = path.replace("${endpointId}", &encode_path(&endpoint_id.to_string()));
        path = path.replace("${deliveryId}", &encode_path(&delivery_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<WebhookDeliveryResponse>(
            Method::GET,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await
    }

    pub async fn create_webhook(
        &self,
        project_id: String,
        body: &CreateWebhookBody,
    ) -> Result<WebhookCreateResponse, SdkError> {
        let mut path = "/v1/projects/${projectId}/webhooks".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<WebhookCreateResponse>(
            Method::POST,
            &path,
            query,
            Some(body),
            headers,
            None,
        ).await
    }

    pub async fn update_webhook(
        &self,
        project_id: String,
        endpoint_id: String,
        body: &UpdateWebhookBody,
    ) -> Result<WebhookResponse, SdkError> {
        let mut path = "/v1/projects/${projectId}/webhooks/${endpointId}".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));
        path = path.replace("${endpointId}", &encode_path(&endpoint_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<WebhookResponse>(
            Method::PATCH,
            &path,
            query,
            Some(body),
            headers,
            None,
        ).await
    }

    pub async fn delete_webhook(
        &self,
        project_id: String,
        endpoint_id: String,
    ) -> Result<(), SdkError> {
        let mut path = "/v1/projects/${projectId}/webhooks/${endpointId}".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));
        path = path.replace("${endpointId}", &encode_path(&endpoint_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute_raw(
            Method::DELETE,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await.map(|_| ())
    }

    pub async fn test_webhook(
        &self,
        project_id: String,
        endpoint_id: String,
    ) -> Result<WebhookDeliveryResponse, SdkError> {
        let mut path = "/v1/projects/${projectId}/webhooks/${endpointId}/test".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));
        path = path.replace("${endpointId}", &encode_path(&endpoint_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<WebhookDeliveryResponse>(
            Method::POST,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await
    }

    pub async fn replay_webhook_delivery(
        &self,
        project_id: String,
        endpoint_id: String,
        delivery_id: String,
    ) -> Result<(), SdkError> {
        let mut path = "/v1/projects/${projectId}/webhooks/${endpointId}/replay/${deliveryId}".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));
        path = path.replace("${endpointId}", &encode_path(&endpoint_id.to_string()));
        path = path.replace("${deliveryId}", &encode_path(&delivery_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute_raw(
            Method::POST,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await.map(|_| ())
    }

}

fn encode_path(value: &str) -> String {
    urlencoding::encode(value).into_owned()
}

fn encode_value<T: Serialize>(value: &T) -> Result<String, SdkError> {
    match serde_json::to_value(value)
        .map_err(|error| SdkError::ValidationError(error.to_string()))?
    {
        serde_json::Value::String(value) => Ok(value),
        value => Ok(value.to_string()),
    }
}

fn encode_cookie_name(name: &str) -> String {
    urlencoding::encode(name).into_owned()
}

fn encode_cookie_value<T: Serialize>(value: &T) -> Result<String, SdkError> {
    let value = encode_value(value)?;
    Ok(urlencoding::encode(&value).into_owned())
}
